### PR TITLE
Add CPUID and MSR snapshot restore test

### PR DESCRIPTION
**Pull request by kalyazin**
## Changes

This adds 2-part tests to verify that CPUID and MSRs retain their values after restoring from a snapshot.
First parts of the tests dump CPUID or MSRs respectively to files and take a snapshot.
Second parts of the tests restore from the snapshot and compare the two CPUIDs or two sets of MSRs values respectively.

## Reason

We need to make sure that when a VM is restored from a snapshot on a different CPU model from the one where the snapshot was taken, CPUID and MSRs hold the same values.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [x] All commits in this PR are signed (`git commit -s`).
- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] New `unsafe` code is documented.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

---

- [ ] This functionality can be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: ../docs/api-change-runbook.md

### CODE REVIEW

File: tests/integration_tests/functional/test_cpu_features.py
Comment (luminitavoicu) [inline tests/integration_tests/functional/test_cpu_features.py:569]: Same here, I would amend these improvements to the commit where you created this function.

File: tests/integration_tests/functional/test_cpu_features.py
Comment (luminitavoicu) [inline tests/integration_tests/functional/test_cpu_features.py:171]: Nit: Can we move this to the beginning of the file for improved readability?

File: tests/integration_tests/functional/test_cpu_features.py
Comment (luminitavoicu) [inline tests/integration_tests/functional/test_cpu_features.py:326]: Nit: same comment as above.

File: tests/integration_tests/functional/test_cpu_features.py
Comment (kalyazin) [reply tests/integration_tests/functional/test_cpu_features.py]: My motivation was the global name needs to be longer to fully convey the context, however it is not practical to use the long name frequently for readability considerations, that's why I introduced an alias.

File: tests/integration_tests/functional/test_cpu_features.py
Comment (kalyazin) [reply tests/integration_tests/functional/test_cpu_features.py]: That's good to know, thanks.

File: tests/integration_tests/functional/test_cpu_features.py
Comment (kalyazin) [reply tests/integration_tests/functional/test_cpu_features.py:680]: as above

File: tests/integration_tests/functional/test_cpu_features.py
Comment (kalyazin) [reply tests/integration_tests/functional/test_cpu_features.py]: Good point. Refactored.

File: resources/tests/msr/wrmsr_list_some.txt
Comment (kalyazin) [reply resources/tests/msr/wrmsr_list_some.txt:1]: That was the idea. I didn't want to disturb any ongoing reviews by rebasing often, but now it looks like a good time to do.

File: tests/integration_tests/functional/test_cpu_features.py
Comment (kalyazin) [reply tests/integration_tests/functional/test_cpu_features.py]: same as above

File: tests/integration_tests/functional/test_cpu_features.py
Comment (kalyazin) [reply tests/integration_tests/functional/test_cpu_features.py:171]: I was thinking about this, but decided to leave it just before the first place where it is used. I can imagine that this file expands with new test functions that don't use the list, and the list definition will lack relevance to the neighbouring code.
Do we have a guideline that advises placing data at the top of the file?

File: tests/integration_tests/functional/test_cpu_features.py
Comment (kalyazin) [reply tests/integration_tests/functional/test_cpu_features.py:326]: same as above

File: tests/integration_tests/functional/test_cpu_features.py
Comment (luminitavoicu) [reply tests/integration_tests/functional/test_cpu_features.py:171]: That makes sense, we can leave this is as. Not a guidance, more like a personal preference.

**Review by luminitavoicu — COMMENTED** [review]
Looks good!

**Review by luminitavoicu — APPROVED** [review]

**Comment by kalyazin** [comment]
I made the following changes compared to the previous version:
- exclude MSR_IA32_TSX_CTRL from the test (limitation is documented in https://github.com/firecracker-microvm/firecracker/pull/3182)
- fix interference between MSR and CPUID tests by using separate artifact directories; they can now be run in the same pipeline
- remove write MSR test with an empty list of MSRs as it is a special case and would complicate the previous item
- rename CPUID tests for consistency

File: tests/integration_tests/functional/test_cpu_features.py
Comment (pb8o) [inline tests/integration_tests/functional/test_cpu_features.py:530]: I think in pandas it is more idiomatic to write `before[~before[].isin()]`, and should get rid of the pylint exception.

File: tests/integration_tests/functional/test_cpu_features.py
Comment (pb8o) [inline tests/integration_tests/functional/test_cpu_features.py:544]: Could be simpler to write it as 2 separate asserts?

File: tests/integration_tests/functional/test_cpu_features.py
Comment (pb8o) [inline tests/integration_tests/functional/test_cpu_features.py:587]: Could use `Path` here and below:

```
tmp_snapshot_artifacts_dir = Path(chroot_dir) / tmp / context.kernel.base_name()
...
mem_fname_in_jail = tmp_snapshot_artifacts_dir / shared_names["mem_fname"]
```

File: tests/integration_tests/functional/test_cpu_features.py
Comment (pb8o) [inline tests/integration_tests/functional/test_cpu_features.py:440]: are all tests in this file x86 specific? We can skip the whole file at the top rather than one by one https://docs.pytest.org/en/7.1.x/reference/reference.html#globalvar-pytestmark

```
pytestmark = pytest.mark.skipif(...)
```

File: tests/integration_tests/functional/test_cpu_features.py
Comment (pb8o) [inline tests/integration_tests/functional/test_cpu_features.py:723]: Would "no_cpu_template" be more descriptive (though longer)?

**Review by mattschlebusch — COMMENTED** [review]
I see you have enough reviewers, so just quick high-level comments:

* The commit "tests: cpu features: share msr exception list" can probably be merged into the same commit that defines the `MSR_EXCEPTION_LIST`.
* The `functional/test_cpu_features.py` file is quite large. A test utilities file with some of the ancilliary functions would be easier and more re-usable.
* As far as I can tell this PR focuses on `T2S` only. While there is a case to be made for deprecating and eventually removing `C3` and `T2`, we should probably be covering those templates as well.

File: tests/integration_tests/functional/test_cpu_features.py
Comment (kalyazin) [reply tests/integration_tests/functional/test_cpu_features.py:723]: Not quite see what you mean. Could you expand?

Comment (pb8o) [reply tests/integration_tests/functional/test_cpu_features.py:723]: When I saw `"none"` I immediately thought it could be a misspelling of Python's `None`, but I think it means `no_cpu_template`. No need to change, more of a nit really.

File: resources/tests/msr/msr_writer.sh
Comment (kalyazin) [reply resources/tests/msr/msr_writer.sh]: done

File: tests/integration_tests/functional/test_cpu_features.py
Comment (kalyazin) [reply tests/integration_tests/functional/test_cpu_features.py]: done

File: tests/integration_tests/functional/test_cpu_features.py
Comment (kalyazin) [reply tests/integration_tests/functional/test_cpu_features.py:541]: This would not be ideal, because we'd like to see both diffs even if the first one fails, to get a bigger picture.

File: tests/integration_tests/functional/test_cpu_features.py
Comment (kalyazin) [reply tests/integration_tests/functional/test_cpu_features.py]: done

File: tests/integration_tests/functional/test_cpu_features.py
Comment (kalyazin) [reply tests/integration_tests/functional/test_cpu_features.py:439]: there is one old test that runs on Arm... I guess we could look into restructuring of this file when user-configurable templates are available and new tests are implemented.

File: tests/integration_tests/functional/test_cpu_features.py
Comment (kalyazin) [reply tests/integration_tests/functional/test_cpu_features.py]: This looks a bit ugly, because the test framework wants None when no template is used, but I need a string to be used as a directory. Not sure how it can be expressed simpler.

**Comment by kalyazin** [comment]
> I see you have enough reviewers, so just quick high-level comments:
> 
> * The commit "tests: cpu features: share msr exception list" can probably be merged into the same commit that defines the `MSR_EXCEPTION_LIST`.
> * The `functional/test_cpu_features.py` file is quite large. A test utilities file with some of the ancilliary functions would be easier and more re-usable.
> * As far as I can tell this PR focuses on `T2S` only. While there is a case to be made for deprecating and eventually removing `C3` and `T2`, we should probably be covering those templates as well.

Thanks, @mattschlebusch .
I squashed the "tests: cpu features: share msr exception list".
I agree, the test file is large, but I refrained from splitting it at this point, as the user-configurable templates are coming and would require some tests. I think it will be more apparent how to refactor it, as more data will be available to extract common code.
I didn't want to test C3 and T2 templates as they are suboptimal with respect to T2S, and we'd have to maintain multiple MSR exception lists, which is in my opinion is not really justified. Customers should be discouraged to use C3 and T2.

**Comment by mattschlebusch** [comment]
> > I see you have enough reviewers, so just quick high-level comments:
> > 
> > * The commit "tests: cpu features: share msr exception list" can probably be merged into the same commit that defines the `MSR_EXCEPTION_LIST`.
> > * The `functional/test_cpu_features.py` file is quite large. A test utilities file with some of the ancilliary functions would be easier and more re-usable.
> > * As far as I can tell this PR focuses on `T2S` only. While there is a case to be made for deprecating and eventually removing `C3` and `T2`, we should probably be covering those templates as well.
> 
> Thanks, @mattschlebusch . I squashed the "tests: cpu features: share msr exception list". I agree, the test file is large, but I refrained from splitting it at this point, as the user-configurable templates are coming and would require some tests. I think it will be more apparent how to refactor it, as more data will be available to extract common code. I didn't want to test C3 and T2 templates as they are suboptimal with respect to T2S, and we'd have to maintain multiple MSR exception lists, which is in my opinion is not really justified. Customers should be discouraged to use C3 and T2.

C3 and T2 haven't been deprecated yet. If they were to be we would need to plan and communicate that. Until then we should be matching the testing for all hardcoded templates. With that said, the testing of T2S added by the PR stands on its own and is not blocked by this point at all.

**Review by luminitavoicu — DISMISSED** [review]
Looks good to me. I only have a couple of nits that I wouldn't block the PR on.

File: resources/tests/msr/wrmsr_list.txt
Comment (kalyazin) [reply resources/tests/msr/wrmsr_list.txt:1]: That is a semi-structured way. I was thinking about documenting the approach I took, but it was rather dirty and incomplete. I will probably create a task to document/enhance it.

File: tests/integration_tests/functional/test_cpu_features.py
Comment (kalyazin) [reply tests/integration_tests/functional/test_cpu_features.py:331]: I did mean to say spans actually :)

File: resources/tests/msr/wrmsr_list.txt
Comment (luminitavoicu) [reply resources/tests/msr/wrmsr_list.txt:1]: That would be a great idea, it would be very useful to have the approach documented.

**Review by pb8o — APPROVED** [review]

**Pull request merged**
